extends Node

export(NodePath) var START_STATE
var states_map = {}

var states_stack = []
var current_state = null
var _active = false setget set_active

func _ready():
	for child in get_children():
		child.connect("finished", self, "_change_state")
	
	initialize(START_STATE)

func initialize(start_state):
	set_active(true)
	states_stack.push_front(get_node(start_state))
	current_state = states_stack[0]
	current_state.enter()

func set_active(value):
	_active = value
	set_physics_process(value)
	set_process_input(value)
	if not _active:
		states_stack = []
		current_state = null

func _input(event):
	current_state.handle_input(event)

func _physics_process(delta):
	current_state.update(delta)

func _change_state(state_name):
	if not _active:
		return
		
	current_state.exit()
	
	if state_name == "previous":
		var previous_state = states_stack.pop_front()
		print("<<<<< " +  previous_state.name)
		if states_stack[0].name == "Move":
			print("=========================================================")
	
	current_state = states_stack[0]


	print(states_stack)

	if state_name != "previous":
		current_state.enter()